tests/func/test_diff.py

```diff
@@ -1,267 +1,274 @@
+        ],
+        "deleted": [],
+        "modified": [
+            {
+                "filename": "file",
+                "checksum": {"old": digest("first"), "new": digest("second")},
+            }
+        ],
+    }
+
+
+def test_deleted(tmp_dir, scm, dvc):
+    tmp_dir.dvc_gen("file", "text", commit="add file")
+    (tmp_dir / "file.dvc").unlink()
+
+    assert dvc.diff() == {
+        "added": [],
+        "deleted": [{"filename": "file", "checksum": digest("text")}],
+        "modified": [],
+    }
+
+
+def test_modified(tmp_dir, scm, dvc):
+    tmp_dir.dvc_gen("file", "first", commit="first version")
+    tmp_dir.dvc_gen("file", "second")
+
+    assert dvc.diff() == {
+        "added": [],
+        "deleted": [],
+        "modified": [
+            {
+                "filename": "file",
+                "checksum": {"old": digest("first"), "new": digest("second")},
+            }
+        ],
+    }
+
+
+def test_refs(tmp_dir, scm, dvc):
+    tmp_dir.dvc_gen("file", "first", commit="first version")
+    tmp_dir.dvc_gen("file", "second", commit="second version")
+    tmp_dir.dvc_gen("file", "third", commit="third version")
+
+    HEAD_2 = digest("first")
+    HEAD_1 = digest("second")
+    HEAD = digest("third")
+
+    assert dvc.diff("HEAD~1") == {
+        "added": [],
+        "deleted": [],
+        "modified": [
+            {"filename": "file", "checksum": {"old": HEAD_1, "new": HEAD}}
+        ],
+    }
+
+    assert dvc.diff("HEAD~2", "HEAD~1") == {
+        "added": [],
+        "deleted": [],
+        "modified": [
+            {"filename": "file", "checksum": {"old": HEAD_2, "new": HEAD_1}}
+        ],
+    }
+
+    with pytest.raises(
+        DvcException, match=r"failed to access revision 'missing'"
+    ):
+        dvc.diff("missing")
+
+
+def test_target(tmp_dir, scm, dvc):
+    tmp_dir.dvc_gen("foo", "foo")
+    tmp_dir.dvc_gen("bar", "bar")
+    scm.add([".gitignore", "foo.dvc", "bar.dvc"])
+    scm.commit("lowercase")
+
+    tmp_dir.dvc_gen("foo", "FOO")
+    tmp_dir.dvc_gen("bar", "BAR")
+    scm.add(["foo.dvc", "bar.dvc"])
+    scm.commit("uppercase")
+
+    assert dvc.diff("HEAD~1", target="foo") == {
+        "added": [],
+        "deleted": [],
+        "modified": [
+            {
+                "filename": "foo",
+                "checksum": {"old": digest("foo"), "new": digest("FOO")},
+            }
+        ],
+    }
+
+    assert dvc.diff("HEAD~1", target="missing") == {
+        "added": [],
+        "deleted": [],
+        "modified": [],
+    }
+
+
+def test_directories(tmp_dir, scm, dvc):
+    tmp_dir.dvc_gen({"dir": {"1": "1", "2": "2"}}, commit="add a directory")
+    tmp_dir.dvc_gen({"dir": {"3": "3"}}, commit="add a file")
+    tmp_dir.dvc_gen({"dir": {"2": "two"}}, commit="modify a file")
+
+    (tmp_dir / "dir" / "2").unlink()
+    dvc.add("dir")
+    scm.add("dir.dvc")
+    scm.commit("delete a file")
+
+    assert dvc.diff(":/init", ":/directory") == {
```

Comment:
Hm, what is ":/init" and ":/directory"? Esp confused about ":/"

Reply:
Those are git references, it searches by commit message:

https://git-scm.com/docs/git-rev-parse#Documentation/git-rev-parse.txt-emlttextgtemegemfixnastybugem

Reply:
I will add a comment on the test.

Reply:
I use them because I think they are more descriptive than the "HEAD~n" syntax

Reply:
Wow, I didn't know that! Neat!